nucliadb/src/migrations/0028_backfill_field_status.py

```diff
@@ -0,0 +1,93 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+
+"""Migration #28
+
+Backfill field status (from error)
+"""
+
+import logging
+from typing import Optional
+
+from nucliadb.common import datamanagers
+from nucliadb.migrator.context import ExecutionContext
+from nucliadb_protos import resources_pb2, writer_pb2
+
+logger = logging.getLogger(__name__)
+
+
+async def migrate(context: ExecutionContext) -> None:
+    start: Optional[str] = ""
+    while True:
+        if start is None:
+            break
+        start = await do_batch(context, start)
+
+
+async def migrate_kb(context: ExecutionContext, kbid: str) -> None: ...
+
+
+async def do_batch(context: ExecutionContext, start: str) -> Optional[str]:
+    async with context.kv_driver.transaction(read_only=False) as txn:
+        async with txn.connection.cursor() as cur:  # type: ignore
+            await cur.execute(
+                """
+                SELECT key, value FROM resources
+                WHERE key ~ '^/kbs/[^/]*/r/[^/]*/f/[^/]*/[^/]*/error$'
+                AND key > %s
+                ORDER BY key
+                LIMIT 100""",
+                (start,),
+            )
+            records = await cur.fetchall()
+            if len(records) == 0:
+                return None
+
+            print(f"Processing {len(records)} field errors from {start}")
+            for key, value in records:
+                error = writer_pb2.Error()
+                error.ParseFromString(value)
+
+                # We do not copy non-DA errors if the resource is not in error.
+                # If we did, we could set some resources that previously were fine to an error status
+                # which is not desired.
+                # There is no way to do this 100% accurate since the /error key is only cleared on field deletion
+                if error.code != writer_pb2.Error.ErrorCode.DATAAUGMENTATION:
+                    parts = key.split("/")
+                    kbid = parts[2]
+                    rid = parts[4]
+                    basic = await datamanagers.resources.get_basic(txn, kbid=kbid, rid=rid)
+                    if basic is None or basic.metadata.status != resources_pb2.Metadata.Status.ERROR:
+                        continue
+
+                # A resource here is in error state, set the field status with the error
+                status = writer_pb2.FieldStatus(
+                    status=writer_pb2.FieldStatus.Status.ERROR,
+                )
+                field_error = writer_pb2.FieldError(
+                    source_error=error,
+                )
+                field_error.created.GetCurrentTime()
+                status.errors.append(field_error)
+                status_key = key[:-5] + "status"
```

Comment:
replace, to be more explicit? 😛

Reply:
Well, replace has the problem that the field may be named "error" and we break it completely. Maybe I can do a split by /, replace last element and join again